barcode/src/lib.rs

```diff
@@ -0,0 +1,1231 @@
+    if let Some(t) = is_char_in_dict(s, &SNV2_DICT) {
+        Err(SerialV2Error::InvalidChar(t))
+    } else {
+        Ok(s)
+    }
+}
+
+/// This type represents a v2 serial number
+#[derive(Debug, PartialEq)]
+pub struct SerialV2<'a>(&'a str);
+
+impl<'a> SerialV2<'a> {
+    /// Get the v2 serial number as a string
+    fn as_str(&'a self) -> &'a str {
+        self.0
+    }
+
+    /// Get the v2 serial number string as a byte slice
+    fn as_bytes(&'a self) -> &'a [u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl<'a> TryFrom<&'a str> for SerialV2<'a> {
+    type Error = SerialV2Error;
+
+    /// Attempt to construct a `SerialV2` instance from a string
+    fn try_from(s: &'a str) -> Result<SerialV2<'a>, Self::Error> {
+        Ok(SerialV2(snv2_fmt_check(s)?))
+    }
+}
+
+/// A type representing the errors that can be encountered while parsing a
+/// serial number string.
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum SerialError {
+    #[error("Part str is the wrong length")]
+    InvalidLength,
+    #[error("Error parsing V1 serial number: {0}")]
+    SerialV1(#[from] SerialV1Error),
+    #[error("Error parsing V2 serial number: {0}")]
+    SerialV2(#[from] SerialV2Error),
+}
+
+/// A data bearing enum wrapping the possible serial number strings
+#[derive(Debug, PartialEq)]
+pub enum Serial<'a> {
+    V1(SerialV1<'a>),
+    V2(SerialV2<'a>),
+}
+
+impl<'a> Serial<'a> {
+    /// Get the serial number as a string
+    pub fn as_str(&'a self) -> &'a str {
+        match self {
+            Serial::V1(s) => s.as_str(),
+            Serial::V2(s) => s.as_str(),
+        }
+    }
+
+    /// Get the serial number string as a byte slice
+    pub fn as_bytes(&'a self) -> &'a [u8] {
+        match self {
+            Serial::V1(s) => s.as_bytes(),
+            Serial::V2(s) => s.as_bytes(),
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a str> for Serial<'a> {
+    type Error = SerialError;
+
+    /// Attempt to construct a `Serial` instance from a string
+    fn try_from(s: &'a str) -> Result<Serial<'a>, Self::Error> {
+        match s.len() {
+            8 => Ok(Serial::V2(SerialV2::try_from(s)?)),
+            11 => Ok(Serial::V1(SerialV1::try_from(s)?)),
+            _ => Err(Self::Error::InvalidLength),
+        }
+    }
+}
+
+/// A type representing the errors that can be encountered while parsing a
+/// barcode string
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum BarcodeError {
+    #[error("The input string has no delimiters")]
+    NoDelim,
+    #[error("The input string has no part number")]
+    NoPartNumber,
+    #[error("The input string has no revision number")]
+    NoRevisionNumber,
+    #[error("The input string has no serial number")]
+    NoSerialNumber,
+    #[error("Barcodes with the 0XV1 or PDV1 prefix must have v1 part numbers")]
+    PartNotV1,
+    #[error("Barcodes with the 0XV2 prefix must have v2 part numbers")]
+    PartNotV2,
+    #[error("Malformed PartError: {0}")]
```

Comment:
We should probably get in the habit of using the `source` attribute in `thiserror` so we can get the full error chain. This would mean deriving [`SlogInlineError`] and following these [instructions](https://github.com/oxidecomputer/slog-error-chain).